From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: willies952002 <admin@domnian.com>
Date: Sun, 3 Jul 2016 18:48:29 -0400
Subject: [PATCH] Add Proxy Mode API

A server can be in one of four modes:
* Bungeecord
* Offline Velocity
* Online Velocity
* None (No Proxy)

diff --git a/src/main/java/com/domnian/paperdragon/ProxyMode.java b/src/main/java/com/domnian/paperdragon/ProxyMode.java
new file mode 100644
index 0000000000000000000000000000000000000000..495fd4f73e4548c99c12e477823c02e7d9c78fc5
--- /dev/null
+++ b/src/main/java/com/domnian/paperdragon/ProxyMode.java
@@ -0,0 +1,47 @@
+package com.domnian.paperdragon;
+
+public enum ProxyMode {
+
+    /**
+     * The server is not running behind any known and supported proxy
+     */
+    NONE,
+
+    /**
+     * The server is running behind a Bungeecord proxy
+     */
+    BUNGEECORD,
+
+    /**
+     * The server is running behind a Velocity proxy, set to offline (unauthenticated) mode
+     */
+    VELOCITY_OFFLINE,
+
+    /**
+     * The server is running behind a Velcity proxy, set to online (authenticated) mode
+     */
+    VELOCITY_ONLINE,
+
+    /**
+     * @deprecated Internal Use Only
+     */
+    @Deprecated
+    UNKNOWN;
+
+    public boolean isProxy() {
+        return this != NONE && this != UNKNOWN;
+    }
+
+    public boolean isOnlineMode() {
+        return this == BUNGEECORD || this == VELOCITY_ONLINE;
+    }
+
+    public boolean isBungee() {
+        return this == BUNGEECORD;
+    }
+
+    public boolean isVelocity() {
+        return this == VELOCITY_OFFLINE || this == VELOCITY_ONLINE;
+    }
+
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index eb0aa2a2a9def0270341ee15d8a7a0629a67a823..90108569c7094b884318d08e495c0eb1d821d12b 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2120,4 +2120,13 @@ public final class Bukkit {
     public static Server.Spigot spigot() {
         return server.spigot();
     }
+
+    // PaperDragon start
+    /**
+     * Get what proxy mode the server is operating in
+     */
+    public static com.domnian.paperdragon.ProxyMode getProxyMode() {
+        return server.getProxyMode();
+    }
+    // PaperDragon end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 967a4bc45de9fdf55d0f65ec7faf2556fb69a6f7..2cc12bfcdd7e32c73b5d8b47f279ab7804a74ddf 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1844,4 +1844,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // PaperDragon start
+    /**
+     * Get what proxy mode the server is operating in
+     */
+    com.domnian.paperdragon.ProxyMode getProxyMode();
+    // PaperDragon end
 }
